### PR TITLE
feat: implement caching metrics

### PR DESCRIPTION
**Pull request by pimeys**
### subgraph entity cache
- hits
- misses
### operation cache:
- hits
- misses
- size

Also enables in-memory operation cache for gateway.

### CODE REVIEW

File: engine/crates/runtime-local/src/operation_cache.rs [hunk 1]
```diff
@@ -49,8 +50,10 @@ impl<V> OperationCache<V> for InMemoryOperationCache<V>
 where
     V: Clone + Send + Sync + 'static + serde::Serialize + serde::de::DeserializeOwned,
 {
-    async fn insert(&self, key: String, value: V) {
+    async fn insert(&self, key: String, value: V) -> u64 {
         self.inner.insert(key, value);
+        self.inner.sync();
```
Comment (tomhoule) [inline engine/crates/runtime-local/src/operation_cache.rs:55]: I wonder if that's expensive. I guess we'll see.

File: gateway/crates/federated-server/src/server/gateway.rs [hunk 2]
```diff
@@ -146,30 +148,36 @@ pub struct GatewayRuntime {
...
         &self.meter
     }
+
     fn hooks(&self) -> &HooksWasi {
         &self.hooks
     }
+
     fn operation_cache_factory(&self) -> &Self::OperationCacheFactory {
-        &()
+        &self.operation_cache_factory
```
Comment (tomhoule) [inline gateway/crates/federated-server/src/server/gateway.rs:180]: Did you manage to clarify why that was a no-op?

**Review by tomhoule — APPROVED** [review]

File: engine/crates/runtime-local/src/operation_cache.rs (hunk 1 above)
Comment (pimeys) [reply engine/crates/runtime-local/src/operation_cache.rs:55]: we run this outside from the main thread anyhow. this is the only way of getting the real counter value right after adding a thing to the cache.

File: gateway/crates/federated-server/src/server/gateway.rs (hunk 2 above)
Comment (pimeys) [reply gateway/crates/federated-server/src/server/gateway.rs:180]: No. There's a PR

https://github.com/grafbase/grafbase/pull/1854

This PR is about adding back the op cache. But it did not do it eventually...

**Pull request merged**